File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/KeyPointCenter/KeyPoint/OneEuroFilter.swift
//
//  OneEuroFilter.swift
//  MetaScripts
//
//  Created by ZhaoTianyu on 2024/7/10.
//

import Foundation

final class OneEuroFilter {
    // Fields
    private var beta: Float
    private var d_cutoff: Vector3
    private var dx_prev = Vector3(x: 0, y: 0, z: 0)
    private var min_cutoff: Vector3
    private var t_prev: Float = 0
    private var x_prev = Vector3(x: 0, y: 0, z: 0)

    // Methods
    init() {
        self.min_cutoff = Vector3(x: 3, y: 3, z: 3)
        self.beta = 0.1
        self.d_cutoff = Vector3(x: 1, y: 1, z: 1)
    }

    init(min_cutoff: Float, beta: Float, d_cutoff: Float) {
        self.min_cutoff = Vector3(x: 3, y: 3, z: 3)
        self.beta = 0.1
        self.d_cutoff = Vector3(x: 1, y: 1, z: 1)
        self.min_cutoff = Vector3(x: min_cutoff, y: min_cutoff, z: min_cutoff)
        self.beta = beta
        self.d_cutoff = Vector3(x: d_cutoff, y: d_cutoff, z: d_cutoff)
    }

    private func AbsVector3(v: Vector3) -> Vector3 {
        return Vector3(x: abs(v.x), y: abs(v.y), z: abs(v.z))
    }

    func correctAndPredict(kp: Vector3, t: Float) -> Vector3 {
        let num = t
        let a = self.smoothing_factor(t_e: num, cutoff: self.d_cutoff)
        let x = (kp - self.x_prev) / num
        let v = self.exponential_smoothing(a: a, x: x, x_prev: self.dx_prev)
        let temp = self.AbsVector3(v: v)
        let cutoff = self.min_cutoff + (temp * self.beta)
        let vector5 = self.smoothing_factor(t_e: num, cutoff: cutoff)
        self.x_prev = kp
        self.dx_prev = v
        self.t_prev = t
        return self.exponential_smoothing(a: vector5, x: kp, x_prev: self.x_prev)
    }

    private func exponential_smoothing(a: Vector3, x: Vector3, x_prev: Vector3) -> Vector3 {
        return Vector3(x: (a.x * x.x) + ((1 - a.x) * x_prev.x),
                       y: (a.y * x.y) + ((1 - a.y) * x_prev.y),
                       z: (a.z * x.z) + ((1 - a.z) * x_prev.z))
    }

    private func smoothing_factor(t_e: Float, cutoff: Vector3) -> Vector3 {
        let vector = (cutoff * 6.283185) * t_e
        return Vector3(x: vector.x / (vector.x + 1),
                       y: vector.y / (vector.y + 1),
                       z: vector.z / (vector.z + 1))
    }
}

File: src/embodied-intelligence/src/3D_human_pose_recognition/ios_sdk/UnityDEBUG/ForDebug/Assets/Scripts/CharacterDriven/oneEuroFilter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

public sealed class OneEuroFilter
{
    // Fields
    private float beta;
    private Vector3 d_cutoff;
    private Vector3 dx_prev;
    private Vector3 min_cutoff;
    private float t_prev;
    private Vector3 x_prev;

    // Methods
    public OneEuroFilter()
    {
        this.min_cutoff = new Vector3(3f, 3f, 3f);
        this.beta = 0.1f;
        this.d_cutoff = new Vector3(1f, 1f, 1f);
    }

    public OneEuroFilter(float min_cutoff, float beta, float d_cutoff)
    {
        this.min_cutoff = new Vector3(3f, 3f, 3f);
        this.beta = 0.1f;
        this.d_cutoff = new Vector3(1f, 1f, 1f);
        this.min_cutoff = new Vector3(min_cutoff, min_cutoff, min_cutoff);
        this.beta = beta;
        this.d_cutoff = new Vector3(d_cutoff, d_cutoff, d_cutoff);
    }

    private Vector3 AbsVector3(Vector3 v)
    {
        return new Vector3(Mathf.Abs(v.x), Mathf.Abs(v.y), Mathf.Abs(v.z));
    }

    public Vector3 CorrectAndPredict(Vector3 kp, float t)
    {
        float num = t;//- this.t_prev;
        Vector3 a = this.smoothing_factor(num, this.d_cutoff);
        Vector3 x = (Vector3)((kp - this.x_prev) / num);
        Vector3 v = this.exponential_smoothing(a, x, this.dx_prev);
        Vector3 cutoff = this.min_cutoff + ((Vector3)(this.beta * this.AbsVector3(v)));
        Vector3 vector5 = this.smoothing_factor(num, cutoff);
        this.x_prev = kp;
        this.dx_prev = v;
        this.t_prev = t;
        return this.exponential_smoothing(vector5, kp, this.x_prev);
    }

    private Vector3 exponential_smoothing(Vector3 a, Vector3 x, Vector3 x_prev)
    {
        return new Vector3((a.x * x.x) + ((1f - a.x) * x_prev.x), (a.y * x.y) + ((1f - a.y) * x_prev.y), (a.z * x.z) + ((1f - a.z) * x_prev.z));
    }

    private Vector3 smoothing_factor(float t_e, Vector3 cutoff)
    {
        Vector3 vector = (Vector3)((6.283185f * cutoff) * t_e);
        return new Vector3(vector.x / (vector.x + 1f), vector.y / (vector.y + 1f), vector.z / (vector.z + 1f));
    }
}


